evap/evaluation/tests/test_models.py

```diff
@@ -719,6 +719,71 @@ def test_disable_cc(self):
+        contributor_both = baker.make(UserProfile)
+
+        # Contributions for evaluation1
+        baker.make(Contribution, evaluation=evaluation1, contributor=responsible1, role=Contribution.Role.CONTRIBUTOR)
+        baker.make(Contribution, evaluation=evaluation1, contributor=editor1, role=Contribution.Role.EDITOR)
+        baker.make(Contribution, evaluation=evaluation1, contributor=contributor1, role=Contribution.Role.CONTRIBUTOR)
+        baker.make(Contribution, evaluation=evaluation1, contributor=contributor_both, role=Contribution.Role.CONTRIBUTOR)
+
+        # Contributions for evaluation2
+        baker.make(
+            Contribution,
+            evaluation=evaluation2,
+            contributor=editor2,
+            role=Contribution.Role.EDITOR
+        )
+        contributor_both_contribution = baker.make(
+            Contribution,
+            evaluation=evaluation2,
+            contributor=contributor_both,
+            role=Contribution.Role.CONTRIBUTOR
+        )
+        contributor2_contribution = baker.make(
+            Contribution,
+            evaluation=evaluation2,
+            contributor=contributor2,
+            role=Contribution.Role.CONTRIBUTOR
+        )
+
+        baker.make(TextAnswer, contribution=contributor_both_contribution)
+        baker.make(TextAnswer, contribution=contributor2_contribution)
+
+        expected_calls = [
+            # these 4 are included since they are contributors for evaluation1 which can publish the average grade
+            call(responsible1, {}, {'user': responsible1, 'evaluations': [evaluation1]}, use_cc=True),
+            call(editor1, {}, {'user': editor1, 'evaluations': [evaluation1]}, use_cc=True),
+            call(contributor1, {}, {'user': contributor1, 'evaluations': [evaluation1]}, use_cc=True),
+            call(contributor_both, {}, {'user': contributor_both, 'evaluations': [evaluation1, evaluation2]}, use_cc=True),
+            # contributor2 has textanswers, so they are notified
+            call(contributor2, {}, {'user': contributor2, 'evaluations': [evaluation2]}, use_cc=True),
+        ]
+
+        with patch('evap.evaluation.models.EmailTemplate.send_to_user') as send_to_user_mock:
+            EmailTemplate.send_contributor_publish_notifications([evaluation1, evaluation2])
+            send_to_user_mock.assert_has_calls(expected_calls, any_order=True)
+
+        # if general textanswers for an evaluation exist, all responsibles should also be notified
+        baker.make(TextAnswer, contribution=evaluation2.general_contribution)
+        expected_calls.append(call(responsible2, {}, {'user': responsible2, 'evaluations': [evaluation2]}, use_cc=True))
+
+        with patch('evap.evaluation.models.EmailTemplate.send_to_user') as send_to_user_mock:
+            EmailTemplate.send_contributor_publish_notifications([evaluation1, evaluation2])
+            send_to_user_mock.assert_has_calls(expected_calls, any_order=True)
```

Comment:
So, I think I _can_ understand this test and it would be fine for me to keep it as-is (there are enough comments to explain what you’re trying to achieve). But, if you want to split this case, I’d suggest to move this part into another case.

Reply:
I think duplicating the setup code is not a good option, so we'd have to refactor the 36 lines of set up out into a different method, but I struggle naming that as it is a very specific scenario that is being set up. For the second method, we would need exactly the same set up with that one additional TextAnswer.

Did you have a specific refactoring in mind that you thought would work quite well?

Reply:
(We wanted to finally get the bug fix done, thus we already merged the current state. I'm still open to refactoring the code, so feel free to reply to this discussion later)